adopt vnet approach from subvending LZ module & modify subnets to work with Azure Policy

kewalaka opened the pull request:
## Description

1) This ports the approach used in the sub vending LZ module

- functionally, this allows a second team to create subnets within a vnet, without terraform overwriting the subnets

2) This also ports the example from @haflidif where subnets are also created using azapi, to allow an immediate association with an NSG & route table, along with aligning it to the AVM way.  As noted, this fixes deployments when faced with typical Azure Policy restrictions.

3) This includes the logic to be able to deploy subnets into an existing vnet (fixes #43)

4) On the examples;

- The 'complete' example is now much more complete.  It deploys all the various resources such as routes, NSGs, DDOS (but disabled, for cost saving), etc.  
- The 'default' example just deploys the minimal vnet
- There is an additional example to test (3)

5) This fixes a number of variables that do not follow AVM spec, noting;

- ``vnet_location`` is now ``location``
- ``virtual_network_dns_servers`` is ``dns_servers``
- ``virtual_network_ddos_protection_plan`` is ``ddos_protection_plan``
- ``virtual_network_address_space`` is ``address_space``
- ``vnet_peering_config`` is ``peerings`` (?)

... and the outputs have been updated to align more closely with AVM too.

6) Change works with recent changes in azapi 1.13

The code passes lint, docs & grept apply checks.

Fixes  #43, #65, #71, #73
Supersedes #62

-->

## Type of Change

<!-- Use the check-boxes [x] on the options that are relevant. -->

- [ ] Non-module change (e.g. CI/CD, documentation, etc.)
- [X] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `locals.version.tf.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `locals.version.tf.json`.
  - [X] Breaking changes and I have bumped the MAJOR version in `locals.version.tf.json`. _edit:_ **bumped the minor version** as we are still < v1.0
  - [X] Update to documentation

# Checklist

- [X] I'm sure there are no other open Pull Requests for the same update/change (except those noted that this supersedes)
- [X] My corresponding pipelines / checks run clean and green without any errors or warnings
- [X] I did run all  [pre-commit](https://azure.github.io/Azure-Verified-Modules/contributing/terraform/terraform-contribution-flow/#5-run-pre-commit-checks) checks

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/terraform -->

kewalaka commented:
@matt-FFFFFF I had to use a 'wait for' due to occasional conflicts when testing, is there might be a better way?

I don't know if this is something that should be noted upstream, it seems like APIs is releasing op locks too early.

haflidif reviewed (commented):
Few comments and some additional approaches, but otherwise very well done implementing subnet using the azapi_provider @kewalaka!

kewalaka commented on variables.tf:
```diff
@@ -57,20 +58,30 @@ If it is set to false, then no telemetry will be collected.
+  })
+  default = null
+}
+
```
this is a convention that has been used elsewhere, i'm not 100% sold on the name either, I'd go with what the consensus suggests.

using a data source isn't recommended in this scenario (ref Matt White's vid - https://www.youtube.com/watch?v=l8k_cOHep28)

haflidif replied:
Yeah totally agree on that Data Source are "evil" within a module and should not be used, but like i said I would recommend using `virtual_network_id` here instead of the `existing_parent_resource` to make it more clear that the module is asking for a virtual network id but of course we check what consensus suggests 💯

kewalaka replied:
yeh that actually does raise a good point - at a minimum it should be suffixed with id - ``existing_parent_resource_id``

@matt-FFFFFF  reckon this is a good one for you to weigh in on with a casting vote, as we should follow this pattern generally, so;

``virtual_network_resource_id`` vs
``existing_parent_resource_id`` or maybe 
``existing_virtual_network_resource_id``
  
 i'm also leaning towards the first option for clarity and being most terraform-like.

kewalaka replied:
thinking on this some more - it would be better to pass in the id rather than the name, that woudl be more 'terraform-like' and align with how other items such as NSGs, route tables, etc are supplied.  Will add this.

kewalaka commented on main.subnet.tf:
```diff
@@ -1,72 +1,58 @@
+    properties = {
+      addressPrefixes = each.value.address_prefixes
+      delegations = each.value.delegations != null ? [
+        for delegation in each.value.delegations : {
```
i think its fair to say that delegations confuses many and the azurerm syntax is.... a little arcane 😅... 

I think i'd like to retain the current object just in case there is reason to keep the name and serviceName separate - re-introducing the action would be additive so not breaking whereas if we change the overall structure that would be.

it feels like there _could_ be a scenario in the future where a subnet is delegated to multiple services.

agreed the for_each is unnecessary

kewalaka replied:
actually i think the for_each is required if keeping the object, because otherwise you're potentially reading subproperties that don't exist.  we can only remove it if going with the simple string for the var approach.  unless i'm missing a trick - i'll post an update soon, i'm just also running avmfix as i noticed that failed too.

kewalaka commented:
as an aside, the current approach uses a data source to grab the subscription details 

``` terraform
data "azurerm_subscription" "this" {}
```

I'm wondering if :

```terraform
data "azurerm_client_config" "this" {}
```

... might be a better choice

kewalaka commented:
I've raised [this ](https://github.com/Azure/terraform-provider-azapi/issues/503) on the AzAPI repo noting the issues with the 409 conflict errors.

I'm thinking also it would be smart to introduce a 'wait for subnet' operations block too , then as a workaround if necessary a caller could call this resource module twice to work around this issue. (unless there's a better way)

haflidif commented:
> I've raised [this ](https://github.com/Azure/terraform-provider-azapi/issues/503) on the AzAPI repo noting the issues with the 409 conflict errors.
> 
> I'm thinking also it would be smart to introduce a 'wait for subnet' operations block too , then as a workaround if necessary a caller could call this resource module twice to work around this issue. (unless there's a better way)

Yeah It's the same issue I stumbled up on during the development of my module, when I was testing the different scenarios on the same vnet but calling my module multiple times in the same vnet, I suspect it's only few seconds that it takes to be ready for each subnet, but enough to hit the operation still in progress error - I would recommend to introduce some time wait block, but not sure how to do that when using a for each loop inside the module to create multiple subnets.

kewalaka commented:
@haflidif hopefully it can be fixed, in the meantime I have added a time wait and an example to the tests that illustrates how to work around it.  Unaware of any sensible ways you can put a pause between 'for_each' objects.  Just about to test it against my own runners.

matt-FFFFFF commented:
See this.

https://github.com/Azure/terraform-provider-azapi/issues/503#issuecomment-2106030842

kewalaka commented:
> See this.
> 
> [Azure/terraform-provider-azapi#503 (comment)](https://github.com/Azure/terraform-provider-azapi/issues/503#issuecomment-2106030842)

onto it.

kewalaka commented:
fixed with no time waits required, pulled them out.  I've converted the vnet peering to azapi too so i can use the same there.  

thanks @matt-FFFFFF ! 

i believe this is ready for review again.

![image](https://github.com/Azure/terraform-azurerm-avm-res-network-virtualnetwork/assets/3146590/96cbaa77-9207-4f3a-9201-a4fcc897d068)

haflidif reviewed (approved):
LGTM 🚀  however I was wondering as this is a significant change from the previous version and introduces quite a breaking change, we should include some information on how to migrate from the last version to the new version - or are we still just in the development phase so we don't need to include any migration guides between versions ? 

This is quite a change between versions, as the module is switching from using azurerm provider to use AzAPI provider.

kewalaka commented:
> LGTM 🚀 
> we should include some information on how to migrate from the last version to the new version
> or are we still just in the development phase so we don't need to include any migration guides between versions ?

That's a question for the AVM core team - I've started with some notes above to call it out - but it is incomplete.  

> This is quite a change between versions, as the module is switching from using azurerm provider to use AzAPI provider.

I don't think he change from AzureRM to AzAPI is so important as that is internal to the module - what's more important is the naming of variables (inputs) and outputs have changed (I believe the changes are all necessary, to align with the spec, or for consistency).  The internal implementation _should_ be changeable without notification as long as the usage is the same - but in this case the usage has changed as well.

haflidif commented:
> > LGTM 🚀
> > we should include some information on how to migrate from the last version to the new version
> > or are we still just in the development phase so we don't need to include any migration guides between versions ?
> 
> That's a question for the AVM core team - I've started with some notes above to call it out - but it is incomplete.
> 
> > This is quite a change between versions, as the module is switching from using azurerm provider to use AzAPI provider.
> 
> I don't think he change from AzureRM to AzAPI is so important as that is internal to the module - what's more important is the naming of variables (inputs) and outputs have changed (I believe the changes are all necessary, to align with the spec, or for consistency). The internal implementation _should_ be changeable without notification as long as the usage is the same - but in this case the usage has changed as well.

Absolutely, I totally agree that normally internal module changes are not required to be notified specially, but in this case here, if you just go from version 0.2.x to 0.3.0 even though you have not change any inputs required for the module, you will get notification on "Destroy" because the azurerm provider is no longer being used to create the virtual networks / subnets hence the state file doesn't match the references. - and then you need to import all existing resources, that alone I think is worth mentioning in breaking changes at least.

kewalaka commented:
Oh good point I forgot to add ``moved`` blocks, will do.

kewalaka commented:
Oh good point I forgot to add ``moved`` blocks, will do.

haflidif commented:
> >[resources will be destroyed and recreated because] the azurerm provider is no longer being used to create the virtual networks / subnets hence the state file doesn't match the references
> 
> 
> 
> _edit_ I had thought that perhaps ``moved`` blocks would work but of course because the resource types are different it won't.  hmm..  I wonder if 'moved' was done from outside the module it would work?
> 
> 
> 
> _edit2_ can't see how moved would help even if called outside the module, looking at the behaviour it just seems to rename the actual resource in the state & then re-run the plan - whereas when look at the state the representation of azapi & azurerm resource is different - so a simple rename won't cut it.
> 
> 
> 
> Don't have a good answer here - any thoughts @matt-FFFFFF  ?

Yeah I think moved block would do the job, nice solution! 👏

jaredfholgate commented on outputs.tf:
```diff
@@ -1,22 +1,22 @@
+output "id" {
```
I think this needs to be `resource_id` now.

kewalaka replied:
@jaredfholgate as an ask - can you please update the terraform output specifics ref back to the output in the shared specs - i was hunting around for where this requirement came from

jaredfholgate replied:
Here it is @kewalaka: https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-rmfr7---category-outputs---minimum-required-outputs

jaredfholgate replied:
Sorry, misread your comment. Will submit a PR.

jaredfholgate commented on variables.tf:
```diff
@@ -12,21 +12,26 @@ The name of the resource group where the resources will be deployed.
+    enable = bool   # (Required) Enable/disable DDoS Protection Plan on Virtual Network.
+  })
+  default     = null
+  description = "AzureNetwork DDoS Protection Plan."
```
This should use HEREDOC and explain the object properties.

jaredfholgate commented on variables.tf:
```diff
@@ -47,6 +52,14 @@ variable "diagnostic_settings" {
+    dns_servers = list(string)
+  })
+  default     = null
+  description = "(Optional) List of IP addresses of DNS servers"
```
Needs better documentation. Objects should be documented in the description.

jaredfholgate commented on variables.tf:
```diff
@@ -57,28 +70,53 @@ If it is set to false, then no telemetry will be collected.
+    id = string
+  })
+  default     = null
+  description = "(Optional) This allows a vnet resource id to be supplied, this allows subnets to be created against an existing vnet."
```
Use HEREDOC and explain the object. It might be better to name it `resource_id`.

jaredfholgate commented on variables.tf:
```diff
@@ -57,28 +70,53 @@ If it is set to false, then no telemetry will be collected.
+    use_remote_gateways                = optional(bool, false)
+  }))
+  default     = {}
+  description = <<DESCRIPTION
```
Need to document the object properties.

jaredfholgate commented on variables.tf:
```diff
@@ -94,100 +132,106 @@ variable "role_assignments" {
-}
-
-variable "tracing_tags_prefix" {
+variable "subscription_id" {
```
What is this for?

kewalaka replied:
Optionally allows vnets to be made in a different subscription.  

Can't always be inferred from existing vnet as this is optional.

If not there this will mean resources are made in the current sub.  

Is this trying to be too clever, maybe.  In theory this module as it stands could drop into to LZ module too.

matt-FFFFFF replied:
See here: this would help: https://github.com/Azure/terraform-provider-azapi/issues/504

kewalaka replied:
my reading is Matt's suggestion is not quite ready to land yet - leaving this open for feedback

jaredfholgate reviewed (changes requested):
Some docs need updating for AVM standards. 

I am unsure why we need to supply the subscription_id as a variable since we are supplying the resource id of the vnet if using an existing one?

kewalaka commented:
> Yeah I think moved block would do the job, nice solution! 👏

Sorry as per my edits I don't think the ``moved`| block will help 😕.  Open to suggestions...

matt-FFFFFF commented:
Can we use import & removed blocks?

kewalaka commented:
> Can we use import & removed blocks?

sounds painful 😅, but ok - i might need some help testing this.  My MSDN subscription has ran out of money until +2 weeks 😿

kewalaka commented:
> Some docs need updating for AVM standards.
> 
> I am unsure why we need to supply the subscription_id as a variable since we are supplying the resource id of the vnet if using an existing one?

explained sub requirement in the code feedback, a copy:

> Optionally this allows vnets to be made in a different subscription.
>
> Can't always be inferred from existing vnet as this is optional.
> 
> If not there this will mean resources are made in the current sub.
>
> Is this trying to be too clever, maybe. In theory this module as it stands could drop into to LZ module too.

haflidif commented:
> > Can we use import & removed blocks?
> 
> 
> 
> sounds painful 😅, but ok - i might need some help testing this.  My MSDN subscription has ran out of money until +2 weeks 😿

I can help with that! I've enough credits on my subscriptions 😊 just tell me what you need ? 💪

kewalaka commented:
> > > Can we use import & removed blocks?
> > 
> > 
> > sounds painful 😅, but ok - i might need some help testing this.  My MSDN subscription has ran out of money until +2 weeks 😿
> 
> I can help with that! I've enough credits on my subscriptions 😊 just tell me what you need ? 💪

Matt is suggesting using recently added TF capability to update the state - using ``removed {}`` and ``import {}``.  These take a source & destination parameter: 

They work the same as terraform state rm and terraform import, except they can be defined in the HCL, which helps in situations like this where we need to shift resources between different resource types.

See: https://developer.hashicorp.com/terraform/language/modules/syntax#removing-modules as an example for "removed".

If it was just refactoring, and the same resource type, then ``moved {}`` would do the trick.

In terms of approach, what I'd suggest is use the new "complete" example from this new PR, because it has more coverage, against the current version of the module, and then the challenge is to make it migrate to this version using the import & removed blocks.

kewalaka commented:
Update; just realised the subnet outputs are in need of love too - I've posted c653c98

jaredfholgate commented:
@kewalaka Given the scale of this change and the likelihood that I am going to add code to it as I test it, I have migrated your branch over to the main repo. Unfortunately, there is no way to change the source branch for a PR, so I'll need to close this PR and open a new one. All your commits are still there with your name on them, so you'll still get the credit for this.

matt-FFFFFF commented:
You'll get credit for the lines of code!

haflidif commented:
Superseded by #106, which consolidated all the example dependency bumps (Azure/avm-utl-regions/azurerm 0.9.0 -> 0.12.0 and Azure/naming/azurerm 0.4.2 -> 0.4.3) into a single change. That PR is now merged, so closing this. Thanks @dependabot! :robot:
